man/bppseqgen.1
.\" SPDX-FileCopyrightText: The Bio++ Development Group
.\"
.\" SPDX-License-Identifier: CECILL-2.1

.TH BPPSEQGEN 1 LOCAL

.SH NAME

bppseqgen - Sequence simulation with Bio++

.SH SYNOPSIS

.B bppseqgen [options]

.SH AVAILABILITY

All UNIX flavors

.SH DESCRIPTION

bppml performs sequence simulation under a given phylogeny or ancestral recombination graph and a large variety of models.

.SH OPTIONS

You should refer to 'info bppsuite' or to the online manual of bppsuite for a complete list of available options.

.TP 5

--noninteractive

generates output for redirection in a file.

.TP

param=file

reads a file for loading options

.SH AUTHOR

The Bio++ Development Team.


ChangeLog
19/03/18 -*- Version 2.4.0 -*-

19/02/18 Julien Dutheil
* Bug fixed in bppPopStats: dN and dS were inverted!

10/12/17 -*- Version 2.3.2 -*-

06/11/17 Julien Dutheil
* Added estimation of kappa to bppPopStats + bugs fixed

06/06/17 -*- Version 2.3.1 -*- 

10/05/17 -*- Version 2.3.0 -*- 

21/12/16 Julien Dutheil
* BppPhySamp migrated to its own package.

17/12/16 Julien Dutheil
* Improved bppPopStats, outgroup analyses, MK table and output to file.

06/10/16 Julien Dutheil
* bppSeqGen now takes Hudson's MS ARG files as input.
* new input.tree.scale option to adjust branch lengths.

22/05/16 Julien Dutheil
* bpppopstats now computes codon statistics per site (e.g. PiN, PiS)
* Added example for bpppopstats.

29/06/15 Julien Dutheil
* New program bppPopStats

28/01/15 Julien Dutheil
* PhySamp now outputs the correct tree!

25/11/14 Julien Dutheil
* bppalnscore output contiguous segments in mase selection.
* bppalnscore site selection works with codon alphabets.

24/11/14 Julien Dutheil
* bppSeqMan can now read alignments and perform site selection priori to any analysis.

28/09/14 -*- Version 2.2.0 -*- 

16/09/14 Julien Dutheil
* bppSeqGen supports generic characters as ancestral states in info file (bug #87).

12/08/14 Julien Dutheil
* bppSeqGen supports simulation under Markov-Modulated models.

06/06/14 Julien Dutheil
* bppPhySamp now output trees as well.

17/04/14 Julien Dutheil
* bppSeqMan: Translate uses global genetic_code argument instead of local 'code' argument.
* All programs use the --seed argument to set random seed for reproduceable results.

08/03/13 -*- Version 0.8.0 -*-

22/01/13 Julien Dutheil
* Compiles with -Wconversion

14/01/13 Mathieu Groussin
* Added COaLA model for proteins in bppML and bppSeqGen.

23/11/12 Laurent Guéguen
* New bppMixedLikelihood program

08/02/12 -*- Version 0.7.0 -*-

21/12/11 Julien Dutheil
* New bppAlnScore program
* Improved bppSeqMan: protein translation now works + more stop codons removal options.

17/06/11 Julien Dutheil
* Improved bppancestor, also allows to compute frequencies for leaves.

11/06/11 -*- Version 0.6.2 -*-
* Small bug fixed + new packages.

03/06/10 -*- Version 0.6.0 -*-
* Compatibility with Bio++ 2.0.0.

03/06/10 Julien Dutheil
* New method in bppPhySamp + new, faster, implementation.

25/03/10 -*- Version 0.5.0 -*-

22/08/09 Julien Dutheil
* Added CMake configuration files, with twick for building the manual (assumes
		texinfo is installed).

22/07/09 Julien Dutheil
* Added program bppTreeDraw.

30/06/09 Julien Dutheil
* BppAncestor: now output expected ancestral states frequencies.
* BppAncestor: new method to compute EB estimates of ancestral states frequencies.

10/06/09 -*- Version 0.4.0 -*-

22/05/09 Julien Dutheil
* Keyval syntax for sequence format description.
* Documentation and example update.
* SeqMan now uses the SequenceApplicationTools method.

15/05/09 Julien Dutheil
* Compatibility update with the libs, new keyval syntax for model
specification.
* Documentation up to date.

02/02/09 Julien Dutheil
* bppConsense now outputs the number of leaves of the input tree.
* New progress bar in bppConsense.

08/01/09 Julien Dutheil
* Added texinfo documentation.
* More bootstrap options in BppDist (the same as in BppML).

25/09/08 -*- Version 0.3.1 -*-

07/12/08 Bastien Boussau & Julien Dutheil
* Bug removed in non-homogeneous models with covarions.

09/10/08 Julien Dutheil
* New action 'invert' in bppSeqMan.

25/09/08 -*- Version 0.3.0 -*-

20/09/08 Julien Dutheil
* New program bppAncestor to reconstruct acestral states and rates.
* New +F models from PhylLib. 

02/07/08 Laurent Guegen
* Bug fixed in bppML: frequencies can be computed from the data with
non-homogeneous models.

19/06/08 Julien Dutheil
* Shell variables are now 'global' in option files.

24/04/08 -*- Version 0.2.3 -*-
* This version requires the CVS version of Bio++ to be built.

03/03/08 Julien Dutheil
* Bug fixed in keep_complete method of BppSeqMan.

27/03/08 Julien Dutheil
* Bug fixed when printing parameters of non-homogeneous models in bppML.

22/01/08 -*- Version 0.2.2 -*-

28/02/08 Celine Scronavacca
* Bug fixed in ReRoot.

22/01/08 -*- Version 0.2.1 -*-

06/02/08 Celine Scornavacca
* Bug fixed in ReRoot + correction of version number in bppSeqGen.

22/01/08 -*- Version 0.2.0 -*-

13/01/08 Julien Dutheil
* More options in bppml: possibility to chose between simple and double
recursion and data compression method.
* Simple recursion method improved.

11/01/08 Julien Dutheil
* Compatibility update
* bppML now output more handy parameter values.

31/12/07 Julien Dutheil
* New program Phylo Sampler.

07/11/07 Julien Dutheil
* Compatibility update with Bio++ cvs.
* Bug fixed in BppSeqGen.
* BppSeqGen can now simulate data from a set of trees (ancestral
  recombination graph).

25/10/07 Julien Dutheil
* Several bug fixed in BppML and BppSeqGen with non-homogeneous models.
* Examples updated + new example option files for non-homogeneous models.

05/10/07 Bastien Boussau & Julien Dutheil
* Non-homogenous sequence simulation in BppSeqGen.
* Model options in files have changed to suit the non-homogeneous case.
  The covarion option now is fully part of the model options, following the
	new syntax 'model=modelname+covarionname', eg 'T92+G2001'.

03/10/07 Julien Dutheil
* New program Bio++ Sequence Manipulator

21/09/07 Julien Dutheil
* Bug fixed in BppDist: branch length are indeed not estimated when parameters
  to ignore contains "BrLen".
* Stop condition when using iterations is now based on topological distance
  instead of likelihood comparisons (phyllib).
* New option optimization.topology_nni.method = [fast|better|phyml]

17/06/07 -*- Version 0.1.0 -*-

16/07/06 Julien Dutheil
* New option bootstrap.verbose for bppML.
* Bug fixed in bootstrap computation.

06/06/06 Julien Dutheil
* New program bppConsense.

05/05/07 Julien Dutheil
* First alpha release of bppsuite.

-*- Old bppML ChangeLog -*-

23/04/07 Julien Dutheil

bppSuite/bppSeqGen.cpp
// SPDX-FileCopyrightText: The Bio++ Development Group
//
// SPDX-License-Identifier: CECILL-2.1

// From the STL:
#include <iostream>
#include <fstream>
#include <iomanip>

using namespace std;

// From bpp-core:
#include <Bpp/Version.h>
#include <Bpp/Numeric/DataTable.h>
#include <Bpp/Text/KeyvalTools.h>
#include <Bpp/App/NumCalcApplicationTools.h>

// From bpp-seq:
#include <Bpp/Seq/SequenceTools.h>
#include <Bpp/Seq/Io/BppOAlignmentWriterFormat.h>

// From bpp-phy:
#include <Bpp/Phyl/App/BppPhylogeneticsApplication.h>
#include <Bpp/Phyl/App/PhylogeneticsApplicationTools.h>
#include <Bpp/Phyl/Simulation/SequenceSimulationTools.h>

#include <Bpp/Phyl/Simulation/EvolutionSequenceSimulator.h>
#include <Bpp/Phyl/Simulation/SimpleSubstitutionProcessSequenceSimulator.h>
#include <Bpp/Phyl/Simulation/GivenDataSubstitutionProcessSequenceSimulator.h>
#include <Bpp/Phyl/Likelihood/PhyloLikelihoods/PhyloLikelihoodContainer.h>
#include <Bpp/Phyl/Likelihood/PhyloLikelihoods/OneProcessSequencePhyloLikelihood.h>
#include <Bpp/Phyl/Likelihood/PhyloLikelihoods/SingleProcessPhyloLikelihood.h>

using namespace bpp;

int main(int args, char** argv)
{
  cout << "******************************************************************" << endl;
  cout << "*            Bio++ Sequence Generator, version " << BPP_VERSION << "             *" << endl;
  cout << "*                                                                *" << endl;
  cout << "* Authors: J. Dutheil                                            *" << endl;
  cout << "*          B. Boussau                       Last Modif. " << BPP_REL_DATE << " *" << endl;
  cout << "*          L. Gueguen                                            *" << endl;
  cout << "*          M. Groussin                                           *" << endl;
  cout << "******************************************************************" << endl;
  cout << endl;

  try
  {
    BppPhylogeneticsApplication bppseqgen(args, argv, "bppseqgen");

    if (args == 1)
    {
      bppseqgen.help("bppSeqGen");
      return 0;
    }

    bppseqgen.startTimer();
    map<string, string> unparsedParams;

    Context context;

    std::shared_ptr<const Alphabet> alphabet = bppseqgen.getAlphabet();

    auto gCode = bppseqgen.getGeneticCode(alphabet);

    // Write to file:
    BppOAlignmentWriterFormat bppoWriter(1);

    ////// Get the optional map of the sequences

    auto mSitesuniq = bppseqgen.getConstAlignmentsMap(alphabet, true, true);

    auto mSites = PhylogeneticsApplicationTools::uniqueToSharedMap<const AlignmentDataInterface>(mSitesuniq);

    /// collection

    std::shared_ptr<SubstitutionProcessCollection> spc = bppseqgen.getCollection(alphabet, gCode, mSites, unparsedParams);

    auto mSeqEvoluniq = bppseqgen.getProcesses(spc, unparsedParams);

    auto mSeqEvol = PhylogeneticsApplicationTools::uniqueToSharedMap<SequenceEvolution>(mSeqEvoluniq);

    /// Get optional phylolikelihoods (in case of posterior simulation)

    auto phyloCont =  bppseqgen.getPhyloLikelihoods(context, mSeqEvol, spc, mSites, "", 0);


    /// ///////////////////
    // Simulators
    vector<string> vSimulName = ApplicationTools::matchingParameters("simul*", bppseqgen.getParams());

    if (vSimulName.size() == 0)
    {
      ApplicationTools::displayWarning("Did not find any descriptor matching `simul*`, so no simulation performed.");
    }

    for (size_t nS = 0; nS < vSimulName.size(); nS++)
    {
      size_t poseq = vSimulName[nS].find("=");
      string suff = vSimulName[nS].substr(5, poseq - 5);

      size_t num = static_cast<size_t>(TextTools::toInt(suff));

      string simulDesc = ApplicationTools::getStringParameter(vSimulName[nS], bppseqgen.getParams(), "", "", true, true);

      map<string, string> argsim;
      string simulName;

      KeyvalTools::parseProcedure(simulDesc, simulName, argsim);

      ApplicationTools::displayMessage("");
      ApplicationTools::displayMessage("Simulation " + TextTools::toString(num));

      //////////////////////

      if (argsim.find("process") == argsim.end() && (argsim.find("phylo") == argsim.end()))
        throw BadIntegerException("bppseqgen. Missing process or phylo argument for simul:", (int)num);


      // Root states
      vector<size_t> states;
      vector<double> rates;
      bool withStates = false;
      bool withRates = false;
      size_t nbSites = 0;

      auto sm = spc->getModel(1)->getStateMap();

      // Data or info at root
      if (argsim.find("data") != argsim.end())
      {
        size_t indData = (size_t)ApplicationTools::getIntParameter("data", argsim, 1, "", true, 0);

        if (mSites.find(indData) == mSites.end())
          throw BadIntegerException("bppseqgen : Unknown data number:", (int)indData);

        auto data = mSites.at(indData);

        auto nseq = RandomTools::giveIntRandomNumberBetweenZeroAndEntry(data->getNumberOfSequences());

        ApplicationTools::displayResult("Using root sequence", data->sequence(nseq).getName());

        string siteSet = ApplicationTools::getStringParameter("input.site.selection", argsim, "none", "", true, 1);

        nbSites = data->getNumberOfSites();

        if (siteSet != "none")
        {
          vector<size_t> vSite;
          try
          {
            vector<int> vSite1 = NumCalcApplicationTools::seqFromString(siteSet, ",", ":");
            for (size_t i = 0; i < vSite1.size(); ++i)
            {
              int x = (vSite1[i] >= 0 ? vSite1[i] : static_cast<int>(nbSites) + vSite1[i]);
              if (x >= 0)
                vSite.push_back(static_cast<size_t>(x));
              else
                throw Exception("bppseqgen. Incorrect negative index for site selection: " + TextTools::toString(x));
            }
          }
          catch (Exception& e)
          {
            string seln;
            map<string, string> selArgs;
            KeyvalTools::parseProcedure(siteSet, seln, selArgs);
            if (seln == "Sample")
            {
              size_t n = ApplicationTools::getParameter<size_t>("n", selArgs, nbSites, "", true, 1);
              bool replace = ApplicationTools::getBooleanParameter("replace", selArgs, false, "", true, 1);

              vSite.resize(n);
              vector<size_t> vPos;
              for (size_t p = 0; p < nbSites; ++p)
              {
                vPos.push_back(p);
              }

              RandomTools::getSample(vPos, vSite, replace);
            }
          }

          nbSites = vSite.size();
        }

        ApplicationTools::displayResult("Number of sites", nbSites);
        states.resize(nbSites);

        withStates = true;

        size_t nbStates = alphabet->getSize();
        std::vector<double> probstate(nbStates);

        auto resChar = alphabet->getResolvedChars();

        for (size_t i = 0; i < nbSites; ++i)
        {
          for (size_t j = 0; j < nbStates; j++)
          {
            probstate[j] = data->getStateValueAt(i, nseq, alphabet->getIntCodeAt(j + 1));
          }

          auto pchar = RandomTools::pickOne<string>(resChar, probstate, true);
          states[i] = RandomTools::pickOne<size_t>(sm->getModelStates(pchar));
        }
      }
      else
      {
        string infosFile = ApplicationTools::getAFilePath("input.infos", argsim, false, true);

        if (infosFile != "none")
        {
          ApplicationTools::displayResult("Site information", infosFile);
          ifstream in(infosFile.c_str());
          auto infos = DataTable::read(in, "\t");
          nbSites = infos->getNumberOfRows();
//      ApplicationTools::displayResult("Number of sites", TextTools::toString(nbSites));
          string rateCol = ApplicationTools::getStringParameter("input.infos.rates", argsim, "pr", "", true, true);
          string stateCol = ApplicationTools::getStringParameter("input.infos.states", argsim, "none", "", true, true);
          withRates = rateCol != "none";
          withStates = stateCol != "none";


          // Specific input data
          if (withRates)
          {
            rates.resize(nbSites);
            vector<string> ratesStrings = infos->getColumn(rateCol);
            for (size_t i = 0; i < nbSites; i++)
            {
              rates[i] = TextTools::toDouble(ratesStrings[i]);
            }
          }
          if (withStates)
          {
            vector<string> ancestralStates = infos->getColumn(stateCol);

            states.resize(nbSites);
            for (size_t i = 0; i < nbSites; i++)
            {
              int alphabetState = alphabet->charToInt(ancestralStates[i]);
              // If a generic character is provided, we pick one state randomly from the possible ones:
              if (alphabet->isUnresolved(alphabetState))
                alphabetState = RandomTools::pickOne<int>(alphabet->getAlias(alphabetState));
              states[i] = RandomTools::pickOne<size_t>(sm->getModelStates(alphabetState));
            }

            string siteSet = ApplicationTools::getStringParameter("input.site.selection", argsim, "none", "", true, 1);

            if (siteSet != "none")
            {
              vector<size_t> vSite;
              try
              {
                vector<int> vSite1 = NumCalcApplicationTools::seqFromString(siteSet, ",", ":");
                for (size_t i = 0; i < vSite1.size(); ++i)
                {
                  int x = (vSite1[i] >= 0 ? vSite1[i] : static_cast<int>(nbSites) + vSite1[i]);
                  if (x >= 0)
                    vSite.push_back(static_cast<size_t>(x));
                  else
                    throw Exception("bppseqgen : incorrect negative index: " + TextTools::toString(x));
                }
              }
              catch (Exception& e)
              {
                string seln;
                map<string, string> selArgs;
                KeyvalTools::parseProcedure(siteSet, seln, selArgs);
                if (seln == "Sample")
                {
                  size_t n = ApplicationTools::getParameter<size_t>("n", selArgs, nbSites, "", true, 1);
                  bool replace = ApplicationTools::getBooleanParameter("replace", selArgs, false, "", true, 1);

                  vSite.resize(n);
                  vector<size_t> vPos;
                  for (size_t p = 0; p < nbSites; ++p)
                  {
                    vPos.push_back(p);
                  }

                  RandomTools::getSample(vPos, vSite, replace);
                }
              }

              nbSites = vSite.size();

              vector<size_t> newStates(nbSites);
              vector<double> newRates(nbSites);

              for (size_t ni = 0; ni < nbSites; ++ni)
              {
                newStates[ni] = states[vSite[ni]];
                newRates[ni]  = rates[vSite[ni]];
              }

              states = newStates;
              rates = newRates;
            }
          }
        }
      }

      ////////////////////////////////////
      /////// Process

      unique_ptr<SequenceSimulatorInterface> ss;

      if (argsim.find("process") != argsim.end())
      {
        size_t indProcess = (size_t)ApplicationTools::getIntParameter("process", argsim, 1, "", true, 0);

        ApplicationTools::displayResult(" Process", TextTools::toString(indProcess));

        if (!spc->hasSubstitutionProcessNumber(indProcess))
        {
          if (mSeqEvol.find(indProcess) == mSeqEvol.end())

            throw BadIntegerException("bppseqgen. Unknown process number:", (int)indProcess);

          ss = make_unique<EvolutionSequenceSimulator>(*mSeqEvol.find(indProcess)->second);
        }
        else
        {
          ss = make_unique<SimpleSubstitutionProcessSequenceSimulator>(spc->getSubstitutionProcess(indProcess));
        }
      }
      else
      {
        size_t indPhylo = (size_t)ApplicationTools::getIntParameter("phylo", argsim, 1, "", true, 0);

        if (!phyloCont)
          throw BadIntegerException("bppseqgen. Empty phylocontainer for simul:", (int)num);

        auto phylo = phyloCont->getPhyloLikelihood(indPhylo);

        ApplicationTools::displayResult(" Phylolikelihood", TextTools::toString(indPhylo));

        if (!phylo)
          throw BadIntegerException("bppseqgen. Unknown phylo number:", (int)indPhylo);

        auto spph = dynamic_pointer_cast<SingleProcessPhyloLikelihood>(phylo);
        auto opsp = dynamic_pointer_cast<OneProcessSequencePhyloLikelihood>(phylo);

        if (!spph && !opsp)
          throw BadIntegerException("bppseqgen. Posterior simulation not implemented for this kind of phylolikelihood. Ask developers.", (int)num);

        std::shared_ptr<LikelihoodCalculationSingleProcess> lcsp = spph ? spph->getLikelihoodCalculationSingleProcess() :
            opsp->getLikelihoodCalculationSingleProcess();

        if (argsim.find("pos") == argsim.end())// Sequence simulation similar to the data, number_of_sites will not be used
          ss = make_unique<GivenDataSubstitutionProcessSequenceSimulator>(lcsp);
        else
        {
          size_t pos = (size_t)ApplicationTools::getIntParameter("pos", argsim, 1, "", true, 0);
          ApplicationTools::displayResult(" Position", TextTools::toString(pos));

          ss = make_unique<SimpleSubstitutionProcessSequenceSimulator>(lcsp, pos);
        }
      }

      // output

      string mfnames = ApplicationTools::getAFilePath("output.sequence.file", argsim, true, false, "", false, "none", true);
      ApplicationTools::displayResult(" Output file", mfnames);

      string mformats = ApplicationTools::getStringParameter("output.sequence.format", argsim, "Fasta", "", false, true);
      ApplicationTools::displayResult(" Output format", mformats);

      auto mintern = ApplicationTools::getBooleanParameter("output.internal.sequences", argsim, false, "", true, 1);
      ApplicationTools::displayBooleanResult(" Output internal", mintern);


      auto gds = dynamic_cast<GivenDataSubstitutionProcessSequenceSimulator*>(ss.get());
      auto pps = dynamic_cast<SubstitutionProcessSequenceSimulator*>(ss.get());

      size_t nbmin = gds ? gds->getNumberOfSites() : pps ? pps->getNumberOfSites() : 100;

      nbSites = (size_t)ApplicationTools::getIntParameter("number_of_sites", argsim, (int)nbmin, "", false, 0);
      ApplicationTools::displayResult(" Number of sites", TextTools::toString(nbSites));


      ss->outputInternalSequences(mintern);
      std::shared_ptr<SiteContainerInterface> sites = 0;

      if (withStates || withRates)
      {
        auto pss = dynamic_cast<SubstitutionProcessSequenceSimulator*>(ss.get());

        if (withStates)
          if (withRates)
            sites = pss ? pss->simulate(rates, states) : SequenceSimulationTools::simulateSites(*ss, rates, states);
          else
            sites = pss ? pss->simulate(states) : SequenceSimulationTools::simulateSites(*ss, states);
        else
          sites = pss ? pss->simulate(rates) : SequenceSimulationTools::simulateSites(*ss, rates);

        ApplicationTools::displayTaskDone();
      }
      else
      {
        ApplicationTools::displayMessage("");
        ApplicationTools::displayTask("Perform simulations");

        sites = ss->simulate(nbSites);
        ApplicationTools::displayTaskDone();
        ApplicationTools::displayMessage("");
      }

      unique_ptr<OAlignment> oAln(bppoWriter.read(mformats));
      // ApplicationTools::displayResult("Output alignment file ", filenames[it.first]);
      // ApplicationTools::displayResult("Output alignment format ", oAln->getFormatName());

      oAln->writeAlignment(mfnames, *sites, true);
    }

    bppseqgen.done();
  }
  catch (exception& e)
  {
    cout << e.what() << endl;
    return 1;
  }

  return 0;
}
